src/io/flutter/samples/FlutterSampleManager.java

```diff
@@ -32,31 +36,112 @@
+import java.nio.file.Files;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.Comparator;
 import java.util.List;
 
 public class FlutterSampleManager {
+  // TODO(pq): remove after testing.
   // See https://github.com/flutter/flutter-intellij/issues/3330.
-  private static final boolean DISABLE_SAMPLES = true;
+  private static final boolean DISABLE_SAMPLES = false;
 
   private static final String SNIPPETS_REMOTE_INDEX_URL = "https://docs.flutter.io/snippets/index.json";
 
   private static final Logger LOG = Logger.getInstance(FlutterSampleManager.class);
+  @NotNull
+  private final FlutterSdk sdk;
 
-  private static List<FlutterSample> SAMPLES;
+  private List<FlutterSample> flutterSamples;
 
-  public static List<FlutterSample> getSamples() {
+  public FlutterSampleManager(@NotNull FlutterSdk sdk) {
+    this.sdk = sdk;
+  }
+
+  public List<FlutterSample> getSamples() {
     if (DISABLE_SAMPLES) {
       return Collections.emptyList();
     }
 
-    if (SAMPLES == null) {
-      // When we're reading from the repo and the file may be changing, consider a fresh read on each access.
-      SAMPLES = loadSamples();
+    if (flutterSamples != null) {
+      return flutterSamples;
+    }
+
+    try {
+      final File tempDir = Files.createTempDirectory("flutter-samples-index").toFile();
+      tempDir.deleteOnExit();
+      final File tempFile = new File(tempDir, "index.json");
+
+      try {
+        final GeneralCommandLine commandLine = sdk.flutterListSamples(tempFile).createGeneralCommandLine(null);
+        LOG.info(commandLine.toString());
+        final OSProcessHandler process = new OSProcessHandler(commandLine);
+        process.addProcessListener(new ProcessAdapter() {
+          @Override
+          public void processTerminated(@NotNull ProcessEvent event) {
+            try {
+              final byte[] bytes = Files.readAllBytes(tempFile.toPath());
+              final String content = new String(bytes);
+              flutterSamples = FlutterSampleManager.readSamples(content);
+            }
+            catch (IOException e) {
+              LOG.warn(e);
+              // On IOException, short circuit checking again.
+              flutterSamples = Collections.emptyList();
+            }
+          }
+        });
+
+        process.startNotify();
+
+        final int timeoutInMs = 2000;
+        // TODO(pq): is this wait right?
+        if (process.waitFor(timeoutInMs)) {
+          LOG.info("Flutter sample listing took more than " + timeoutInMs + "ms");
+          // Don't cache here so that we can try again later.
+        }
+      }
+      catch (ExecutionException e) {
+        LOG.warn(e);
+        // On ExecutionException, short circuit checking again.
+        flutterSamples = Collections.emptyList();
+      }
     }
-    return SAMPLES;
+    catch (IOException e) {
+      LOG.warn(e);
+      // On IOException, short circuit checking again.
+      flutterSamples = Collections.emptyList();
```

Comment:
If the network is unreliable, will the IDE have to be restarted to reset flutterSamples?

Reply:
I'm going to look into this some more.  There are a few wrinkles.

Thanks!